### ISSUE TITLE
Not able to set a locale or redirect to specific locale from the server in a middleware or plugin.

### ISSUE THREAD

**Issue by dennisadriaans** [nuxt-modules/i18n#2554]
### Environment

- Operating System: Linux
- Node Version:     v18.15.0
- Nuxt Version:     3.7.4
- CLI Version:      3.9.1
- Nitro Version:    2.7.0
- Package Manager:  npm@9.5.0
- Builder:          -
- User Config:      devtools, app, modules, i18n, runtimeConfig
- Runtime Modules:  @nuxtjs/i18n@8.0.0-rc.5, @pinia/nuxt@0.5.1
- Build Modules:    -

### Reproduction

Install Nuxt i18n, try to call setLocale in Nuxt plugin our middleware.

### Describe the bug

It seems like I cannot set a locale or redirect to specific locale from the server in a middleware or plugin.

`$i18n.switchLocalePath()`
`$i18n.setLocale( localeCookie.value )`

Nitro error:
[nitro] [unhandledRejection] Error: [nuxt] A composable that requires access to the Nuxt instance was called outside of a plugin.

Tested in both middleware and plugin.

**Comment by dennisadriaans** [comment]
I am unable to proceed with testing my actual bug due to the apparent malfunction of the basic language switcher.

Please refer to my attempted reproduction steps.

https://stackblitz.com/edit/github-m2j2vu?file=nuxt.config.ts

**Comment by BobbieGoede** [comment]
Your reproduction doesn't have any pages, so there's no page for `switchLocalePath` to return. I modified your reproduction [here](https://stackblitz.com/edit/github-m2j2vu-rothxs?file=pages%2Findex.vue) to demonstrate, if there are no pages `vue-router` won't be included as a dependency by Nuxt (see [docs](https://nuxt.com/docs/guide/directory-structure/app#minimal-usage)).

**Comment by edgarsn** [comment]
Same happens for me when calling setLocale in middleware like this:

```typescript
export default defineNuxtRouteMiddleware(async (to) => {
    const {$i18n} = useNuxtApp();
    const userLocaleFromBackend = 'en';
    
    await $i18n.setLocale(userLocaleFromBackend);
});
```
Using Nuxt 3.9.0.

**8.0.0-rc.7** ✅ (Works fine for me)
**8.0.0-rc.8** ✅ (Works fine for me)
**8.0.0-rc.9** ❌ (Starts to fail)
**8.0.0-rc.10** ❌
**8.0.0-rc.11** ❌
**8.0.0** ❌

**Comment by ucw** [comment]
setting `experimental: {asyncContext: true}` is a possible solution

**Comment by michaelzangl** [comment]
I took some time to investigate the issue.

The problem is that `setLocale` asynchronously calls `detectRedirect` which uses the `useNuxtApp` hook.  So calling it in anything on server that triggers the async code can cause problems (e.g. waiting for an API call and then setting the locale based on the result of that call).

**Comment by BobbieGoede** [comment]
It's still unclear what the expected behaviour is, and without a minimal reproduction I can only guess what's going on.

If you're trying to change the locale in a route middleware, you're best off using a redirect from the middleware using `switchLocalePath`.

**Comment by ucw** [comment]
https://stackblitz.com/edit/nuxt-starter-i64ji9?file=plugins%2Fi18n.ts reproduction from my use-case. I need to make a custom locale detection. I expect that calling the setLocale method from the plugin will change the language, not throw an error.

**Comment by BobbieGoede** [comment]
@ucw 
Thanks for the reproduction, I think I know what's causing this issue and how to fix it (#2777 didn't fully resolve this). Will likely have a fix ready this week.

**Comment by BobbieGoede** [comment]
@ucw 
The issue in your reproduction should be resolved in the latest edge release, I modified your reproduction to demonstrate [here](https://stackblitz.com/edit/nuxt-starter-t9cykm?file=package.json,plugins%2Fi18n.ts). 

Can you try installing the edge version in your project and confirm if this resolves the issue for you?

**Comment by ucw** [comment]
@BobbieGoede I checked my project with edge version, everything works fine. Thank you for your work!

**Comment by BobbieGoede** [comment]
@ucw that's great to hear! I just published v8.1.1 which includes these changes. 

I'll close this issue now, if anyone else is still experiencing this or a similar problem, please open a new issue with a (minimal) reproduction.
